pkg/reconciler/knativeserving/knativeserving.go

```diff
@@ -166,20 +182,67 @@ func (r *Reconciler) ensureFinalizerRemoval(_ context.Context, _ *mf.Manifest, i
 	return nil
 }
 
-// Delete obsolete resources from previous versions
-func (r *Reconciler) deleteObsoleteResources(ctx context.Context, manifest *mf.Manifest, instance *servingv1alpha1.KnativeServing) error {
-	resources := []*unstructured.Unstructured{
-		// istio-system resources from 0.3.
-		common.NamespacedResource("v1", "Service", "istio-system", "knative-ingressgateway"),
-		common.NamespacedResource("apps/v1", "Deployment", "istio-system", "knative-ingressgateway"),
-		common.NamespacedResource("autoscaling/v1", "HorizontalPodAutoscaler", "istio-system", "knative-ingressgateway"),
-		// config-controller from 0.5
-		common.NamespacedResource("v1", "ConfigMap", instance.GetNamespace(), "config-controller"),
-	}
-	for _, r := range resources {
-		if err := manifest.Client.Delete(r); err != nil {
-			return err
-		}
+func (r *Reconciler) retrieveManifest(ctx context.Context, version string, instance *servingv1alpha1.KnativeServing) (mf.Manifest, error) {
+	if val, found := r.manifests[version]; found {
+		return val, nil
 	}
-	return nil
+
+	logger := logging.FromContext(ctx)
+	koDataDir := os.Getenv("KO_DATA_PATH")
+	manifesrDir := fmt.Sprintf("knative-serving/%s", version)
+	manifest, err := mf.NewManifest(filepath.Join(koDataDir, manifesrDir),
+		mf.UseClient(r.mfClient),
+		mf.UseLogger(zapr.NewLogger(logger.Desugar()).WithName("manifestival")))
+
+	if err != nil {
+		return manifest, err
+	}
+
+	if len(manifest.Resources()) == 0 {
+		return manifest, fmt.Errorf("unable to find the manifest for the Knative Serving version %s", version)
+	}
+
+	// Transform the manifest
+	transformers, err := r.transform(ctx, instance)
+	if err != nil {
+		return mf.Manifest{}, err
+	}
+
+	manifestTransformed, err := manifest.Transform(transformers...)
+	if err != nil {
+		return mf.Manifest{}, err
+	}
```

Comment:
Nope! You can't cache transformed manifests, as they should change when the KS CR changes. We must perform the transform for each reconcile in the `getTarget` and `getCurrent` functions.

Reply:
Done.

Reply:
Not committed yet?